fix(helm): Port accept dependency in requirements.yaml from charts directory

hickeyma opened the pull request:
This is a port of #6578 from Helm v2 with modifications to integrate with v3 infrastructure.

**What this PR does / why we need it**:

**Special notes for your reviewer**:
For manual testing, use steps similar to the following:
- Create chart `foo`
- Create chart `sub-foo`
- Move `sub-foo` to `foo/charts`
- Add the following dependencies to `foo/Chart.yaml`:
```console
dependencies:
- name: sub-foo
  version: 0.1.0
- name: mysql
  version: "1.3.2"
  repository: "https://kubernetes-charts.storage.googleapis.com"
```
- Do a dependency update: `helm dep up foo`

It should fail similar to the following without the PR:
```console
$ helm dep up foo
Error: no repository definition for . Please add them via 'helm repo add'
Note that repositories must be URLs or aliases. For example, to refer to the "example"
repository, use "https://charts.example.com/" or "@example" instead of
"example". Don't forget to add the repo, too ('helm repo add').
```

PR should fix it similar to the following:
```console
$ helm dep update foo
Hang tight while we grab the latest from your chart repositories...
..Successfully got an update from the "stable" chart repository
Update Complete. ⎈Happy Helming!⎈
Saving 2 charts
Dependency sub-foo did not declare a repository. Assuming it exists in the charts directory
Downloading mysql from repo https://kubernetes-charts.storage.googleapis.com
Deleting outdated charts

$ ls -lrt foo/charts/
total 20
drwxr-xr-x 4 usr1 usr1  4096 Oct  8 20:02 sub-foo
-rw-rw-r-- 1 usr1 usr1  2739 Oct  8 20:05 sub-foo-0.1.0.tgz
-rw-r--r-- 1 usr1 usr1 10635 Oct  8 20:37 mysql-1.3.2.tgz
```
**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

hickeyma committed:
Port #6578 to Helm 3

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

thomastaylor312 commented on internal/resolver/resolver_test.go:
```diff
@@ -85,6 +85,29 @@ func TestResolve(t *testing.T) {
+			},
+		},
+		{
+			name: "repo from valid path under charts path",
```
I think this test name needs to be "repo from _invalid_ path under charts path" correct?

hickeyma replied:
Sure. Done.

thomastaylor312 reviewed (changes requested):
Looking pretty good and unit tests seem to cover things. Just a small change needed in a unit test name for clarity

hickeyma committed:
Update after reviw

Review comments:
- https://github.com/helm/helm/pull/6611#discussion_r332745703

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thanks for review @thomastaylor312. Ready again for review.
